src/neml2/models/solid_mechanics/AssociativeJ2Normality.cxx

```diff
@@ -0,0 +1,100 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "neml2/models/solid_mechanics/AssociativeJ2Normality.h"
+#include "neml2/tensors/SSR4.h"
+
+using vecstr = std::vector<std::string>;
+
+namespace neml2
+{
+register_NEML2_object(AssociativeJ2Normality);
+
+OptionSet
+AssociativeJ2Normality::expected_options()
+{
+  OptionSet options = Model::expected_options();
+  options.set<LabeledAxisAccessor>("trial_cauchy_stress") = vecstr{"state", "S"};
+  options.set<LabeledAxisAccessor>("plastic_flow_direction") = vecstr{"forces", "NM"};
+  options.set<LabeledAxisAccessor>("isotropic_hardening_direction");
+  options.set<LabeledAxisAccessor>("kinematic_hardening_direction");
+  return options;
+}
+
+AssociativeJ2Normality::AssociativeJ2Normality(const OptionSet & options)
+  : Model(options),
+    trial_cauchy_stress(
+        declare_input_variable<SR2>(options.get<LabeledAxisAccessor>("trial_cauchy_stress"))),
+    plastic_flow_direction(
+        declare_output_variable<SR2>(options.get<LabeledAxisAccessor>("plastic_flow_direction"))),
+    isotropic_hardening_direction(
+        options.get<LabeledAxisAccessor>("isotropic_hardening_direction")),
+    kinematic_hardening_direction(options.get<LabeledAxisAccessor>("kinematic_hardening_direction"))
+{
+  if (!isotropic_hardening_direction.empty())
+    declare_output_variable<Scalar>(isotropic_hardening_direction);
+  if (!kinematic_hardening_direction.empty())
+    declare_output_variable<SR2>(kinematic_hardening_direction);
+
+  setup();
+}
+
+void
+AssociativeJ2Normality::set_value(const LabeledVector & in,
+                                  LabeledVector * out,
+                                  LabeledMatrix * dout_din,
+                                  LabeledTensor3D * d2out_din2) const
+{
+  neml_assert_dbg(!d2out_din2, "AssociativeJ2Normality doesn't implement second derivatives.");
+
+  const auto options = in.options();
+
+  auto S = in.get<SR2>(trial_cauchy_stress);
+  auto devS = S.dev();
+  auto NM = devS / devS.norm(EPS);
```

Comment:
Can't we just get this from the derivative of the yield surface with respect to stress (evaluated at the trial state)?

Reply:
true